Denon MC4000 mapping

timrae opened the pull request:
Here's my mapping for the [Denon DJ MC4000](http://denondj.com/products/view/mc4000) controller. Please refer to the [documentation on the wiki](http://mixxx.org/wiki/doku.php/denon_mc4000).

It's basically complete now, with a few TODOs:
1. The "pad mode" button for the hot cues doesn't currently map to anything; i.e. it's permanently configured to trigger hot cues 1 - 4. In Serato it's possible to switch these to cues 5 - 8, configure them as cue loops, or cue rolls. It might be nice for someone to implement this.
2. It would be nice if the library encoder push button could switch between selecting songs / vs. crates / vs/ playlists, etc. I'll probably send a PR for that soonish when I figure out the best way to do it.
3. The panel / back buttons should also be updated similarly to above point after PR
4. The FX beats encoder doesn't map to anything
5. The autloop button doesn't work that well. I'll try send a PR to fix that up and adjust here accordingly.
6. The BPM tap button appears to be mapped correctly but not do anything (should check source code)
7. I sent an email to Denon DJ asking about the MIDI message to get the current controller states but still no reply after a week. According to my reading on the net it's possible to sniff the midi port on mac, so maybe I'll take the controller to a friends house or something.

sblaisot commented:
please add file removal from windows uninstaller
https://mixxx.org/wiki/doku.php/contributing_mappings#windows_installer_update

Be-ing commented on res/controllers/Denon-MC4000-scripts.js:
```diff
@@ -0,0 +1,544 @@
+        }
+    }    
+    // Pitch slider rate change, MSB
+    this.rateMsb = function(midichan, control, value, status, group) {
```
Does this work with XML using the fourteen-bit-lsb/fourteen-bit-msb options?

Be-ing commented on res/controllers/Denon-MC4000-scripts.js:
```diff
@@ -0,0 +1,544 @@
+        engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
+    };
+    // Sync toggle button
+    // TODO: shift + sync should be used to diable sync instead of using toggle
```
What do you mean? Pressing the sync button when sync is enabled disables sync. You could use shift + sync for another control. Mapping it to quantize may work well.

timrae replied:
I think the way it works in Serato is that when you press the sync button while the sync function is enabled, the sync is recalculated at the current time point. You need to use shift + sync to turn off the sync function entirely. I'm not sure how or if I should map that same functionality to Mixxx.

Be-ing replied:
That's not how sync works in Mixxx. When the rate of a deck is changed that has sync enabled, all decks with sync enabled change rate by the same amount. Pressing sync again removes a deck from having its rate tied to others with sync enabled.

Be-ing commented on res/controllers/Denon-MC4000-scripts.js:
```diff
@@ -0,0 +1,544 @@
+    this.currentRateIndex = 0;
+    
+    // Set default rate range (to +/- 8% like Serato and Technics)
+    engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
```
Having the script set this upon initialization may frustrate users who have chosen a different rate range in Mixxx's preferences.

timrae replied:
Can we access the preference value from JS?

ywwg replied:
not yet! But we're working on that functionality

Be-ing replied:
Does `engine.getValue(this.group, "rateRange")` work?

rryan replied:
@Be-ing is right here, you can access the user's preferred rate range via the "rateRange" control.

timrae replied:
Alright, I guess I could hardcode the options listed in preferences into
the JS and toggle through that.
On 7 May 2016 10:59, "RJ Ryan" <notifications@github.com> wrote:

> In res/controllers/Denon-MC4000-scripts.js
> <https://github.com/mixxxdj/mixxx/pull/938#discussion_r62409476>:
>
> > +
> > +MC4000.OnVuMeterChangeR = function(value, group, control) {
> > +    midi.sendShortMsg(0xBF, 0x45, value*0x7F);
> > +}
> > +
> > +MC4000.Deck = function (channel) {
> > +    // Some state variables
> > +    this.isVinylMode = false;
> > +    this.rateMsbValue = 0;
> > +    this.midiChannel = channel;
> > +    this.engineChannel = channel + 1;
> > +    this.group = '[Channel' + this.engineChannel + ']';
> > +    this.currentRateIndex = 0;
> > +
> > +    // Set default rate range (to +/- 8% like Serato and Technics)
> > +    engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
>
> @Be-ing <https://github.com/Be-ing> is right here, you can access the
> user's preferred rate range via the "rateRange" control.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/938/files/4dcee7f172408d704ffea276057901605bb8d7b5#r62409476>
>

Be-ing commented on res/controllers/Denon-MC4000-scripts.js:
```diff
@@ -0,0 +1,544 @@
+    // Shift + x 1/2
+    // x2
+    // Shift + x2
+    // Reloop
```
I've made a JS object for this for the P32 mapping the other day. https://github.com/Be-ing/mixxx/blob/4dbfa1d5235e54156c2a2b79fccd08410944f694/res/controllers/Hercules-P32-scripts.js#L254

Be-ing commented:
Leave a comment whenever you've added more to be reviewed. Are you still having trouble logging into the wiki?

timrae commented:
@Be-ing 

Yes, I'm going to play around with moving more stuff to XML, I'll let you know when I'm ready for final review.

> Are you still having trouble logging into the wiki?

Yes, the wiki is still broken. I've tried clearing my browser cache and everything, I think it's a problem on the server end that should be easily reproducible by trying to log into a new account.

timrae commented:
Ok I think this is ready for re-review, I've taken out a lot of the javascript as it was unnecessary. I've also added mappings for most of the previously unmapped features, and added documentation to the wiki; please refer to the updated PR description above.

With regards to the pitch fader range preference, I don't think this is a big deal to be honest. The mapping currently matches the Serato behavior, so I think it's reasonable. If a feature is added to access the preferences through JS then we can look at changing it to toggle through that list instead.

Let me know if you're happy with everything and I'll squash down to a single commit before you merge.

timrae commented:
Although,  I could probably do the hot cue buttons in pure XML as well, and get rid of all of the JS boiler plate code from @Be-ing.

This JS code was not as useful in practice as it originally appeared to be... I think better documentation of the XML input / outputs format (adding concrete examples for the common controls) would remove the confusion there, and that may be a better solution that directing people to a JavaScript library.

timrae commented:
Okkkkkkkkkkk I'm pretty happy with this now, I think it's ready for merging.

Be-ing commented:
I went through the wiki and it mostly looks good. I edited the language of the wiki to make it more concise and explain what is meant by "hardware only".

As for the pad modes, I'm not sure what could be a good use for them. There already are separate controls for loops and effects. Maybe providing access to more hotcues would be the best way to map them. Anyone have other ideas?

The effect section mapping seems odd. It seems like a waste to have the left and right effects sections control exactly the same values. I think it would make more sense to map each side to one EffectUnit and use the enable buttons for effects 1-3 within that unit. There isn't a very good answer as for what to do with the knobs because Mixxx lacks a way of control individual effects with a single knob (see https://bugs.launchpad.net/mixxx/+bug/1575973 ). They could be mapped to the parameters of the first effect in the EffectUnit with the "beats" encoder controlling dry/wet. Shift + beats could control the super knob.

timrae commented:
In the wiki you added the note:

> Other ranges for the fader can be selected in Mixxx'​s Preferences in the Interface section.

This isn't actually true. The mapping will initially respect the user's pitch range setting, but pressing the button will then gracefully switch between the standard 8/16/50 ranges, as you can see by looking at the code. This is the best I'm willing to do  at the moment.

timrae commented:
> The effect section mapping seems odd. It seems like a waste to have the left and right effects sections control exactly the same values. I think it would make more sense to map each side to one EffectUnit and use the enable buttons to enable effects 1-3 within that unit for the corresponding deck.

But there are only 4 EffectUnits in Mixxx and 6 on the MC4000....?

Be-ing commented:
Does this make it clear for the wiki? "Other ranges for the fader can be selected in Mixxx's Preferences in the Interface section, but pressing this button with Shift will only cycle through 8%, 16%, and 50%."

Be-ing commented:
> But there are only 4 EffectUnits in Mixxx and 6 on the MC4000 right....?

Each EffectUnit has several effects with in it. What I am proposing is to map the left effect controls to [EffectRack1_EffectUnit1] and the right effect controls to [EffectRack1_EffectUnit2]. Upon initialization, the script could set [EffectRack1_EffectUnit1], group_[Channel1]_enable to 1 and [EffectRack1_EffectUnit2], group_[Channel2]_enable to 1. Users could then select which effects within those EffectUnits are enabled with the effect on buttons of the controller. See http://mixxx.org/wiki/doku.php/effects_framework#controls

Does that make sense?

Be-ing commented:
@timrae : ping. Thoughts about the effects mapping?

timrae commented:
Sorry I didn't really understand your proposal to be honest and got quite
busy with work... I'm keen to get this merged though, so maybe I can find
some time next week if you can describe your proposed effect mapping a bit
more concretely.

On 27 Nov. 2016 01:40, "Be" <notifications@github.com> wrote:

> @timrae <https://github.com/timrae> : ping. Thoughts about the effects
> mapping?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/938#issuecomment-263072946>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4qXwPcxE5SFuO8cV1mOIYJSslpknks5rCGDkgaJpZM4IVKwy>
> .
>

Be-ing commented:
I see now that I was confused about how you've mapped the effects because of this sentence on the wiki: "Note: Both decks control the same FX unit." But when I looked at the XML, they both control effect units 1-3. I think it is a waste to use the knobs on both sides for exactly the same controls. Perhaps on one side you could keep the knobs mapped to the mix knob of effect units 1-3 and on the other side map the knobs to the superknob of effect units 1-3?

timrae commented:
I think that would be a pretty good idea for my own personal mapping, but I feel everyday users might find that a bit surprising / unintuitive. As it is, it works approximately the same as Serato, which is nice IMHO ([POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)). 

How about just keeping it how it is and adding a note in the wiki on suggested tweaks to get the most of the out of the controller? Or having two separate basic / advanced mappings...?

Be-ing commented:
Either way is unintuitive. I think it'd be better to go with the more useful of the two (one side as dry/wet, one side as superknobs). Alternatively, if you want to put a bit more work into it, you could map each side to one effect unit, map the knobs to control individual effect parameters, and implement a way to switch between effects within the effects unit. That is what I did for the [Hercules P32](http://mixxx.org/wiki/doku.php/hercules_p32_dj).

timrae commented:
Alright, I'll try get onto it this week

timrae commented:
@Be-ing
It seems some of the FX (e.g. flanger & reverb, the only ones I use) don't have super knobs defined, which kinda sucks...

timrae commented:
Also, most of the FX have more than 3 knobs, which makes your second suggestion a bit awkward... It's also a lot more work.

Be-ing commented:
> It seems some of the FX (e.g. flanger & reverb, the only ones I use) don't have super knobs defined, which kinda sucks...

Yeah, [that is an issue](https://bugs.launchpad.net/mixxx/+bug/1466966). IMO this should be fixed for 2.1. I'll play around with superknob linkings for all the effects and try to come up with more sensible defaults. If you want to take a try at proposing some more default mappings or giving feedback on what I come up with, that would be appreciate. You can edit superknob linkings manually in Deere and Shade.

> Also, most of the FX have more than 3 knobs, which makes your second suggestion a bit awkward... It's also a lot more work.

Yeah, it's not perfect, it's just a way to work with a system lacking a better way. It is a lot of work, and it probably wouldn't be all that intuitive to use anyway (it works well for the P32 because I had a whole layer of a 4x4 pad grid that I didn't know what else to do with), so if you don't want to implement it, that's fine.

timrae committed:
Add MC4000 mapping

timrae commented:
Alright, if we can get superknob linkings for all the FX in v 2.1 then that's gonna be the easiest way. I've updated the PR to use the right side for the super knob.

Be-ing commented:
Cool, LGTM! 👍 I updated the wiki to reflect the change. Can someone merge this?

Thanks for reminding me about the lack of default superknob linkings. Fixing that will benefit more than just the mapping for this controller.

Be-ing commented:
Thank you @timrae ! Mixxx has been lacking in mappings for recent mid-grade controllers. Now there's a good option for that :)
